app/com/arpnetworking/metrics/portal/alerts/impl/DatabaseAlertRepository.java

```diff
@@ -0,0 +1,365 @@
+                    .setOperator(ebeanAlert.getOperator())
+                    .setPeriod(Period.seconds(ebeanAlert.getPeriod()))
+                    .setService(ebeanAlert.getService())
+                    .setStatistic(ebeanAlert.getStatistic())
+                    .setValue(new DefaultQuantity.Builder()
+                            .setValue(ebeanAlert.getQuantityValue())
+                            .setUnit(ebeanAlert.getQuantityUnit())
+                            .build())
+                    .setExtensions(mergeExtensions(ebeanAlert.getNagiosExtension()))
+                    .build();
+            values.add(result);
+        });
+
+        // Transform the results
+        return new DefaultQueryResult<>(values, pagedAlerts.getTotalRowCount(), etag.toString());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getAlertCount() {
+        assertIsOpen();
+        return Ebean.find(models.ebean.Alert.class)
+                .findRowCount();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addOrUpdateAlert(final Alert alert) {
+        assertIsOpen();
+        LOGGER.debug()
+                .setMessage("Upserting alert")
+                .addData("alert", alert)
+                .log();
+
+
+        try (Transaction transaction = Ebean.beginTransaction()) {
+            models.ebean.Alert ebeanAlert = Ebean.find(models.ebean.Alert.class)
+                    .where()
+                    .eq("uuid", alert.getId())
+                    .findUnique();
+            boolean created = false;
+            if (ebeanAlert == null) {
+                ebeanAlert = new models.ebean.Alert();
+                created = true;
+            }
+
+            ebeanAlert.setCluster(alert.getCluster());
+            ebeanAlert.setUuid(alert.getId());
+            ebeanAlert.setMetric(alert.getMetric());
+            ebeanAlert.setContext(alert.getContext());
+            ebeanAlert.setNagiosExtension(getNagiosExtensionFromMap(alert.getExtensions()));
+            ebeanAlert.setName(alert.getName());
+            ebeanAlert.setOperator(alert.getOperator());
+            ebeanAlert.setPeriod(alert.getPeriod().toStandardSeconds().getSeconds());
+            ebeanAlert.setQuantityValue(alert.getValue().getValue());
+            ebeanAlert.setQuantityUnit(alert.getValue().getUnit().isPresent() ? alert.getValue().getUnit().get() : null);
+            ebeanAlert.setStatistic(alert.getStatistic());
+            ebeanAlert.setService(alert.getService());
+            _alertQueryGenerator.saveAlert(ebeanAlert);
+            transaction.commit();
+
+            LOGGER.info()
+                    .setMessage("Upserted alert")
+                    .addData("alert", alert)
+                    .addData("isCreated", created)
+                    .log();
+        } catch (final IOException e) {
+            LOGGER.error()
+                    .setMessage("Failed to close the transaction")
```

Comment:
The IOException here is a checked exception thrown by the autocloseable transaction object. So in this case, we know that the commit went through. We dont know if the commits were made in case of the other exception, like in case of an exception after the transaction.commit() statement. If the caller knows whether or not the queries went through, it will help show appropriate error messages to user. Right?